File: src/main/java/pl/swmud/ns/swaedit/gui/WelcomeScreen.java
package pl.swmud.ns.swaedit.gui;

import com.trolltech.qt.core.QRect;
import com.trolltech.qt.core.QTimer;
import com.trolltech.qt.core.Qt;
import com.trolltech.qt.gui.QApplication;
import com.trolltech.qt.gui.QBrush;
import com.trolltech.qt.gui.QCloseEvent;
import com.trolltech.qt.gui.QColor;
import com.trolltech.qt.gui.QImage;
import com.trolltech.qt.gui.QMouseEvent;
import com.trolltech.qt.gui.QPaintEvent;
import com.trolltech.qt.gui.QPainter;
import com.trolltech.qt.gui.QWidget;

public class WelcomeScreen extends QWidget {
    
    private QImage img;
    private QRect screen;
    Signal1<Boolean> closed = new Signal1<Boolean>();
    
    public WelcomeScreen() {
        super(SWAEdit.ref);
        setWindowTitle(SWAEdit.ref.windowTitle());
        setWindowIcon(QApplication.windowIcon());
        img = new QImage("images/header.png");
        screen = QApplication.desktop().screenGeometry();
        setFixedSize(screen.size());
        setAttribute(Qt.WidgetAttribute.WA_DeleteOnClose);
        setWindowModality(Qt.WindowModality.ApplicationModal);
        setWindowFlags(Qt.WindowType.Dialog);
        setWindowState(Qt.WindowState.WindowFullScreen);
        closed.connect(SWAEdit.ref, "splashScreen_closed()");
    }
    
    public void showNow() {
        if (!img.isNull()) {
            QTimer.singleShot(10000, this, "close()");
            showFullScreen();
        }
        else {
            closed.emit(true);
        }
    }
    
    public void paintEvent(QPaintEvent e) {
        super.paintEvent(e);
        QPainter p = new QPainter(this);
        p.fillRect(screen, new QBrush(QColor.black));
        p.drawImage(screen.size().width()/2-img.size().width()/2,screen.size().height()/2-img.size().height()/2,img);
    }
    
    public void mousePressEvent(QMouseEvent e) {
        close();
    }
    
    public void closeEvent(QCloseEvent e) {
        hide();
        closed.emit(true);
        super.closeEvent(e);
    }
}

File: src/gui/WelcomeScreen.cpp
#include "gui/WelcomeScreen.h"

#include <QApplication>
#include <QCloseEvent>
#include <QCoreApplication>
#include <QMouseEvent>
#include <QPainter>
#include <QPaintEvent>
#include <QScreen>
#include <QTimer>

WelcomeScreen::WelcomeScreen(QWidget *parent)
    : QWidget(parent)
{
    if (parent) {
        setWindowTitle(parent->windowTitle());
    }
    setWindowIcon(QApplication::windowIcon());

    img_ = QImage(QCoreApplication::applicationDirPath() + "/images/header.png");

    setAttribute(Qt::WA_DeleteOnClose);
    setWindowModality(Qt::ApplicationModal);
    setWindowFlags(Qt::Dialog);
}

void WelcomeScreen::showNow()
{
    if (!img_.isNull()) {
        QTimer::singleShot(10000, this, &WelcomeScreen::close);
        showFullScreen();
    } else {
        emit closed();
    }
}

void WelcomeScreen::paintEvent(QPaintEvent *event)
{
    QWidget::paintEvent(event);

    QPainter p(this);
    p.fillRect(rect(), Qt::black);

    if (!img_.isNull()) {
        int x = width() / 2 - img_.width() / 2;
        int y = height() / 2 - img_.height() / 2;
        p.drawImage(x, y, img_);
    }
}

void WelcomeScreen::mousePressEvent(QMouseEvent *event)
{
    Q_UNUSED(event);
    close();
}

void WelcomeScreen::closeEvent(QCloseEvent *event)
{
    hide();
    emit closed();
    QWidget::closeEvent(event);
}
